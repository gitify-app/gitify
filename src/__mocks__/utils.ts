export function setPlatform(platform: NodeJS.Platform) {
  Object.defineProperty(process, 'platform', {
    value: platform,
  });
}
