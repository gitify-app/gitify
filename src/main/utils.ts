import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { dialog, shell } from 'electron';
import log from 'electron-log';
import type { Menubar } from 'menubar';

export function takeScreenshot(mb: Menubar) {
  const date = new Date();
  const dateStr = date.toISOString().replace(/:/g, '-');

  const capturedPicFilePath = `${os.homedir()}/${dateStr}-gitify-screenshot.png`;
  mb.window.capturePage().then((img) => {
    fs.writeFile(capturedPicFilePath, img.toPNG(), () =>
      log.info(`Screenshot saved ${capturedPicFilePath}`),
    );
  });
}

export function resetApp(mb: Menubar) {
  const cancelButtonId = 0;
  const resetButtonId = 1;

  const response = dialog.showMessageBoxSync(mb.window, {
    type: 'warning',
    title: 'Reset Gitify',
    message:
      'Are you sure you want to reset Gitify? You will be logged out of all accounts',
    buttons: ['Cancel', 'Reset'],
    defaultId: cancelButtonId,
    cancelId: cancelButtonId,
  });

  if (response === resetButtonId) {
    mb.window.webContents.send('gitify:reset-app');
    mb.app.quit();
  }
}

export function openLogsDirectory() {
  const logDirectory = path.dirname(log.transports.file?.getFile()?.path);

  if (!logDirectory) {
    log.error('Could not find log directory!');
    return;
  }

  shell.openPath(logDirectory);
}
