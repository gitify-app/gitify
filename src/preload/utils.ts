import { ipcRenderer } from 'electron';

import type { EventData, EventType } from '../shared/events';

/**
 * Send renderer event without expecting a response
 * @param event the type of event to send
 * @param data the data to send with the event
 */
export function sendMainEvent(event: EventType, data?: EventData): void {
  ipcRenderer.send(event, data);
}

/**
 * Send renderer event and expect a response
 * @param event the type of event to send
 * @param data the data to send with the event
 * @returns
 */
export function invokeMainEvent(
  event: EventType,
  data?: string,
): Promise<string> {
  return ipcRenderer.invoke(event, data);
}

/**
 * Handle renderer event without expecting a response
 */
export function onRendererEvent(
  event: EventType,
  listener: (event: Electron.IpcRendererEvent, args: string) => void,
) {
  ipcRenderer.on(event, listener);
}
