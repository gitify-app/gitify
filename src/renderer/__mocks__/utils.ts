import * as helpers from '../utils/helpers';

export function mockDirectoryPath() {
  jest.spyOn(helpers, 'getDirectoryPath').mockReturnValue('/mocked/dir/name');
}

/**
 * Ensure stable snapshots for our randomized emoji use-cases
 */
export function ensureStableEmojis() {
  global.Math.random = jest.fn(() => 0.1);
}
