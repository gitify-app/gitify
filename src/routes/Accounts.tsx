import {
  ArrowLeftIcon,
  KeyIcon,
  PersonIcon,
  PlusIcon,
  SignOutIcon,
} from '@primer/octicons-react';

import { type FC, useCallback, useContext } from 'react';
import { useNavigate } from 'react-router-dom';

import { AppContext } from '../context/App';

import { AuthMethodIcon } from '../components/icons/AuthMethodIcon';
import { PlatformIcon } from '../components/icons/PlatformIcon';
import type { Account } from '../types';
import { getAccountUUID, getDeveloperSettingsURL } from '../utils/auth/utils';
import {
  openExternalLink,
  updateTrayIcon,
  updateTrayTitle,
} from '../utils/comms';
import {
  isOAuthAppLoggedIn,
  isPersonalAccessTokenLoggedIn,
} from '../utils/helpers';

export const AccountsRoute: FC = () => {
  const { auth, logoutFromAccount } = useContext(AppContext);
  const navigate = useNavigate();

  const logoutAccount = useCallback((account: Account) => {
    logoutFromAccount(account);
    navigate(-1);
    updateTrayIcon();
    updateTrayTitle();
  }, []);

  const openProfile = (account: Account) => {
    const url = new URL(`https://${account.hostname}`);
    url.pathname = account.user.login;
    openExternalLink(url.toString());
  };

  const openHost = (hostname: string) => {
    openExternalLink(`https://${hostname}`);
  };

  const openDeveloperSettings = (account: Account) => {
    const url = getDeveloperSettingsURL(account);
    openExternalLink(url);
  };

  const loginWithPersonalAccessToken = useCallback(() => {
    return navigate('/login-personal-access-token', { replace: true });
  }, []);

  const loginWithOAuthApp = useCallback(() => {
    return navigate('/login-oauth-app', { replace: true });
  }, []);

  const buttonClass =
    'hover:text-gray-500 py-1 px-2 my-1 mx-2 focus:outline-none';

  return (
    <div
      className="flex flex-1 flex-col h-screen dark:bg-gray-dark dark:text-white"
      data-testid="accounts"
    >
      <div className="flex justify-between items-center mt-2 py-2 mx-8">
        <button
          type="button"
          className="focus:outline-none"
          title="Go Back"
          onClick={() => navigate(-1)}
        >
          <ArrowLeftIcon
            size={20}
            className="hover:text-gray-400"
            aria-label="Go Back"
          />
        </button>

        <h3 className="text-lg font-semibold">Accounts</h3>
      </div>

      <div className="flex-grow overflow-x-auto px-8">
        <div className="flex flex-col mt-4 text-sm">
          {auth.accounts.map((account) => (
            <div
              key={getAccountUUID(account)}
              className="flex justify-between items-center bg-gray-100 dark:bg-gray-900 rounded-md p-1 mb-4"
            >
              <div className="ml-2 text-xs">
                <button
                  type="button"
                  className="cursor-pointer font-semibold mb-1 text-sm"
                  title="Open Profile"
                  onClick={() => openProfile(account)}
                >
                  @{account.user.login}
                  <span
                    hidden={!account.user?.name}
                    className="pl-1 font-medium text-xs italic"
                  >
                    - {account.user?.name}
                  </span>
                </button>

                <button
                  type="button"
                  className="cursor-pointer mb-1 ml-1 align-middle"
                  title="Open Host"
                  onClick={() => openHost(account.hostname)}
                >
                  <PlatformIcon type={account.platform} size={12} />
                  {account.platform} - {account.hostname}
                </button>
                <button
                  type="button"
                  className="cursor-pointer ml-1 align-middle"
                  title="Open Developer Settings"
                  onClick={() => openDeveloperSettings(account)}
                >
                  <AuthMethodIcon type={account.method} size={12} />
                  {account.method}
                </button>
              </div>
              <div>
                <button
                  type="button"
                  className={buttonClass}
                  title={`Logout ${account.user.login}`}
                  onClick={() => logoutAccount(account)}
                >
                  <SignOutIcon
                    size={20}
                    aria-label={`Logout ${account.user.login}`}
                  />
                </button>
              </div>
            </div>
          ))}
        </div>
      </div>

      <div className="flex justify-between items-center bg-gray-200 dark:bg-gray-darker py-1 px-8 text-sm">
        <div className="font-semibold italic">Add new account</div>
        <div>
          <button
            type="button"
            className={buttonClass}
            title="Login with Personal Access Token"
            onClick={loginWithPersonalAccessToken}
            hidden={isPersonalAccessTokenLoggedIn(auth)}
          >
            <KeyIcon size={18} aria-label="Login with Personal Access Token" />
            <PlusIcon size={10} className="ml-1 mb-2" />
          </button>
          <button
            type="button"
            className={buttonClass}
            title="Login with OAuth App"
            onClick={loginWithOAuthApp}
            hidden={isOAuthAppLoggedIn(auth)}
          >
            <PersonIcon size={20} aria-label="Login with OAuth App" />
            <PlusIcon size={10} className="mb-2" />
          </button>
        </div>
      </div>
    </div>
  );
};
