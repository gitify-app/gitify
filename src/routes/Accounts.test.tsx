import { act, fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import {
  mockAuth,
  mockGitHubAppAccount,
  mockOAuthAccount,
  mockPersonalAccessTokenAccount,
  mockSettings,
} from '../__mocks__/state-mocks';
import { AppContext } from '../context/App';
import * as comms from '../utils/comms';
import * as links from '../utils/links';

import { AccountsRoute } from './Accounts';

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

describe('routes/Accounts.tsx', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('General', () => {
    it('should render itself & its children', async () => {
      await act(async () => {
        render(
          <AppContext.Provider
            value={{
              auth: {
                accounts: [
                  mockPersonalAccessTokenAccount,
                  mockOAuthAccount,
                  mockGitHubAppAccount,
                ],
              },
              settings: mockSettings,
            }}
          >
            <MemoryRouter>
              <AccountsRoute />
            </MemoryRouter>
          </AppContext.Provider>,
        );
      });

      expect(screen.getByTestId('accounts')).toMatchSnapshot();
    });

    it('should go back by pressing the icon', async () => {
      await act(async () => {
        render(
          <AppContext.Provider
            value={{
              auth: mockAuth,
              settings: mockSettings,
            }}
          >
            <MemoryRouter>
              <AccountsRoute />
            </MemoryRouter>
          </AppContext.Provider>,
        );
      });

      fireEvent.click(screen.getByLabelText('Go Back'));
      expect(mockNavigate).toHaveBeenNthCalledWith(1, -1);
    });
  });

  describe('Account interactions', () => {
    it('open profile in external browser', async () => {
      const openAccountProfileMock = jest.spyOn(links, 'openAccountProfile');

      await act(async () => {
        render(
          <AppContext.Provider
            value={{
              auth: {
                accounts: [mockPersonalAccessTokenAccount],
              },
              settings: mockSettings,
            }}
          >
            <MemoryRouter>
              <AccountsRoute />
            </MemoryRouter>
          </AppContext.Provider>,
        );
      });

      fireEvent.click(screen.getByTitle('Open Profile'));

      expect(openAccountProfileMock).toHaveBeenCalledTimes(1);
      expect(openAccountProfileMock).toHaveBeenCalledWith(
        mockPersonalAccessTokenAccount,
      );
    });

    it('open host in external browser', async () => {
      const openExternalLinkMock = jest.spyOn(comms, 'openExternalLink');

      await act(async () => {
        render(
          <AppContext.Provider
            value={{
              auth: {
                accounts: [mockPersonalAccessTokenAccount],
              },
              settings: mockSettings,
            }}
          >
            <MemoryRouter>
              <AccountsRoute />
            </MemoryRouter>
          </AppContext.Provider>,
        );
      });

      fireEvent.click(screen.getByTitle('Open Host'));

      expect(openExternalLinkMock).toHaveBeenCalledTimes(1);
      expect(openExternalLinkMock).toHaveBeenCalledWith('https://github.com');
    });

    it('open developer settings in external browser', async () => {
      const openExternalLinkMock = jest.spyOn(comms, 'openExternalLink');

      await act(async () => {
        render(
          <AppContext.Provider
            value={{
              auth: {
                accounts: [mockPersonalAccessTokenAccount],
              },
              settings: mockSettings,
            }}
          >
            <MemoryRouter>
              <AccountsRoute />
            </MemoryRouter>
          </AppContext.Provider>,
        );
      });

      fireEvent.click(screen.getByTitle('Open Developer Settings'));

      expect(openExternalLinkMock).toHaveBeenCalledTimes(1);
      expect(openExternalLinkMock).toHaveBeenCalledWith(
        'https://github.com/settings/tokens',
      );
    });

    it('should logout', async () => {
      const logoutFromAccountMock = jest.fn();
      const updateTrayIconMock = jest.spyOn(comms, 'updateTrayIcon');
      const updateTrayTitleMock = jest.spyOn(comms, 'updateTrayTitle');

      await act(async () => {
        render(
          <AppContext.Provider
            value={{
              auth: {
                accounts: [mockPersonalAccessTokenAccount],
              },
              settings: mockSettings,
              logoutFromAccount: logoutFromAccountMock,
            }}
          >
            <MemoryRouter>
              <AccountsRoute />
            </MemoryRouter>
          </AppContext.Provider>,
        );
      });

      fireEvent.click(screen.getByTitle('Logout octocat'));

      expect(logoutFromAccountMock).toHaveBeenCalledTimes(1);

      expect(updateTrayIconMock).toHaveBeenCalledTimes(1);
      expect(updateTrayIconMock).toHaveBeenCalledWith();
      expect(updateTrayTitleMock).toHaveBeenCalledTimes(1);
      expect(updateTrayTitleMock).toHaveBeenCalledWith();
      expect(mockNavigate).toHaveBeenNthCalledWith(1, -1);
    });
  });

  describe('Add new accounts', () => {
    it('should show login with personal access token', async () => {
      await act(async () => {
        render(
          <AppContext.Provider
            value={{
              auth: { accounts: [mockOAuthAccount] },
              settings: mockSettings,
            }}
          >
            <MemoryRouter>
              <AccountsRoute />
            </MemoryRouter>
          </AppContext.Provider>,
        );
      });

      expect(screen.getByTitle('Login with Personal Access Token').hidden).toBe(
        false,
      );

      fireEvent.click(screen.getByTitle('Login with Personal Access Token'));
      expect(mockNavigate).toHaveBeenNthCalledWith(
        1,
        '/login-personal-access-token',
        {
          replace: true,
        },
      );
    });

    it('should show login with oauth app', async () => {
      await act(async () => {
        render(
          <AppContext.Provider
            value={{
              auth: { accounts: [mockPersonalAccessTokenAccount] },
              settings: mockSettings,
            }}
          >
            <MemoryRouter>
              <AccountsRoute />
            </MemoryRouter>
          </AppContext.Provider>,
        );
      });

      expect(screen.getByTitle('Login with OAuth App').hidden).toBe(false);

      fireEvent.click(screen.getByTitle('Login with OAuth App'));
      expect(mockNavigate).toHaveBeenNthCalledWith(1, '/login-oauth-app', {
        replace: true,
      });
    });
  });
});
