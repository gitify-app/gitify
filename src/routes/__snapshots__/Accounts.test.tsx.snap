// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes/Accounts.tsx General should render itself & its children 1`] = `
<div
  class="flex flex-1 flex-col h-screen dark:bg-gray-dark dark:text-white"
  data-testid="accounts"
>
  <div
    class="flex justify-between items-center mt-2 py-2 mx-8"
  >
    <button
      class="focus:outline-none"
      title="Go Back"
      type="button"
    >
      <svg
        aria-label="Go Back"
        class="hover:text-gray-400"
        fill="currentColor"
        focusable="false"
        height="20"
        role="img"
        style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
        viewBox="0 0 16 16"
        width="20"
      >
        <path
          d="M7.78 12.53a.75.75 0 0 1-1.06 0L2.47 8.28a.75.75 0 0 1 0-1.06l4.25-4.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042L4.81 7h7.44a.75.75 0 0 1 0 1.5H4.81l2.97 2.97a.75.75 0 0 1 0 1.06Z"
        />
      </svg>
    </button>
    <h3
      class="text-lg font-semibold"
    >
      Accounts
    </h3>
  </div>
  <div
    class="flex-grow overflow-x-auto px-8"
  >
    <div
      class="flex flex-col mt-4 text-sm"
    >
      <div
        class="flex justify-between items-center bg-gray-100 dark:bg-gray-900 rounded-md p-1 mb-4"
      >
        <div
          class="ml-2 text-xs"
        >
          <button
            class="cursor-pointer font-semibold mb-1 text-sm"
            title="Open Profile"
            type="button"
          >
            @
            octocat
            <span
              class="pl-1 font-medium text-xs italic"
            >
              - 
              Mona Lisa Octocat
            </span>
          </button>
          <button
            class="cursor-pointer mb-1 ml-1 align-middle"
            title="Open Host"
            type="button"
          >
            <span
              aria-label="GitHub Cloud"
              class="mr-1"
            >
              <svg
                aria-hidden="true"
                class="octicon octicon-mark-github"
                fill="currentColor"
                focusable="false"
                height="12"
                style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
                viewBox="0 0 16 16"
                width="12"
              >
                <path
                  d="M8 0c4.42 0 8 3.58 8 8a8.013 8.013 0 0 1-5.45 7.59c-.4.08-.55-.17-.55-.38 0-.27.01-1.13.01-2.2 0-.75-.25-1.23-.54-1.48 1.78-.2 3.65-.88 3.65-3.95 0-.88-.31-1.59-.82-2.15.08-.2.36-1.02-.08-2.12 0 0-.67-.22-2.2.82-.64-.18-1.32-.27-2-.27-.68 0-1.36.09-2 .27-1.53-1.03-2.2-.82-2.2-.82-.44 1.1-.16 1.92-.08 2.12-.51.56-.82 1.28-.82 2.15 0 3.06 1.86 3.75 3.64 3.95-.23.2-.44.55-.51 1.07-.46.21-1.61.55-2.33-.66-.15-.24-.6-.83-1.23-.82-.67.01-.27.38.01.53.34.19.73.9.82 1.13.16.45.68 1.31 2.69.94 0 .67.01 1.3.01 1.49 0 .21-.15.45-.55.38A7.995 7.995 0 0 1 0 8c0-4.42 3.58-8 8-8Z"
                />
              </svg>
            </span>
            GitHub Cloud
             - 
            github.com
          </button>
          <button
            class="cursor-pointer ml-1 align-middle"
            title="Open Developer Settings"
            type="button"
          >
            <span
              aria-label="Personal Access Token"
              class="mr-1"
            >
              <svg
                aria-hidden="true"
                class="octicon octicon-key"
                fill="currentColor"
                focusable="false"
                height="12"
                style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
                viewBox="0 0 16 16"
                width="12"
              >
                <path
                  d="M10.5 0a5.499 5.499 0 1 1-1.288 10.848l-.932.932a.749.749 0 0 1-.53.22H7v.75a.749.749 0 0 1-.22.53l-.5.5a.749.749 0 0 1-.53.22H5v.75a.749.749 0 0 1-.22.53l-.5.5a.749.749 0 0 1-.53.22h-2A1.75 1.75 0 0 1 0 14.25v-2c0-.199.079-.389.22-.53l4.932-4.932A5.5 5.5 0 0 1 10.5 0Zm-4 5.5c-.001.431.069.86.205 1.269a.75.75 0 0 1-.181.768L1.5 12.56v1.69c0 .138.112.25.25.25h1.69l.06-.06v-1.19a.75.75 0 0 1 .75-.75h1.19l.06-.06v-1.19a.75.75 0 0 1 .75-.75h1.19l1.023-1.025a.75.75 0 0 1 .768-.18A4 4 0 1 0 6.5 5.5ZM11 6a1 1 0 1 1 0-2 1 1 0 0 1 0 2Z"
                />
              </svg>
            </span>
            Personal Access Token
          </button>
        </div>
        <div>
          <button
            class="hover:text-gray-500 py-1 px-2 my-1 mx-2 focus:outline-none"
            title="Logout octocat"
            type="button"
          >
            <svg
              aria-label="Logout octocat"
              class="octicon octicon-sign-out"
              fill="currentColor"
              focusable="false"
              height="20"
              role="img"
              style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
              viewBox="0 0 16 16"
              width="20"
            >
              <path
                d="M2 2.75C2 1.784 2.784 1 3.75 1h2.5a.75.75 0 0 1 0 1.5h-2.5a.25.25 0 0 0-.25.25v10.5c0 .138.112.25.25.25h2.5a.75.75 0 0 1 0 1.5h-2.5A1.75 1.75 0 0 1 2 13.25Zm10.44 4.5-1.97-1.97a.749.749 0 0 1 .326-1.275.749.749 0 0 1 .734.215l3.25 3.25a.75.75 0 0 1 0 1.06l-3.25 3.25a.749.749 0 0 1-1.275-.326.749.749 0 0 1 .215-.734l1.97-1.97H6.75a.75.75 0 0 1 0-1.5Z"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="flex justify-between items-center bg-gray-100 dark:bg-gray-900 rounded-md p-1 mb-4"
      >
        <div
          class="ml-2 text-xs"
        >
          <button
            class="cursor-pointer font-semibold mb-1 text-sm"
            title="Open Profile"
            type="button"
          >
            @
            octocat
            <span
              class="pl-1 font-medium text-xs italic"
            >
              - 
              Mona Lisa Octocat
            </span>
          </button>
          <button
            class="cursor-pointer mb-1 ml-1 align-middle"
            title="Open Host"
            type="button"
          >
            <span
              aria-label="GitHub Enterprise Server"
              class="mr-1"
            >
              <svg
                aria-hidden="true"
                class="octicon octicon-server"
                fill="currentColor"
                focusable="false"
                height="12"
                style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
                viewBox="0 0 16 16"
                width="12"
              >
                <path
                  d="M1.75 1h12.5c.966 0 1.75.784 1.75 1.75v4c0 .372-.116.717-.314 1 .198.283.314.628.314 1v4a1.75 1.75 0 0 1-1.75 1.75H1.75A1.75 1.75 0 0 1 0 12.75v-4c0-.358.109-.707.314-1a1.739 1.739 0 0 1-.314-1v-4C0 1.784.784 1 1.75 1ZM1.5 2.75v4c0 .138.112.25.25.25h12.5a.25.25 0 0 0 .25-.25v-4a.25.25 0 0 0-.25-.25H1.75a.25.25 0 0 0-.25.25Zm.25 5.75a.25.25 0 0 0-.25.25v4c0 .138.112.25.25.25h12.5a.25.25 0 0 0 .25-.25v-4a.25.25 0 0 0-.25-.25ZM7 4.75A.75.75 0 0 1 7.75 4h4.5a.75.75 0 0 1 0 1.5h-4.5A.75.75 0 0 1 7 4.75ZM7.75 10h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM3 4.75A.75.75 0 0 1 3.75 4h.5a.75.75 0 0 1 0 1.5h-.5A.75.75 0 0 1 3 4.75ZM3.75 10h.5a.75.75 0 0 1 0 1.5h-.5a.75.75 0 0 1 0-1.5Z"
                />
              </svg>
            </span>
            GitHub Enterprise Server
             - 
            github.gitify.io
          </button>
          <button
            class="cursor-pointer ml-1 align-middle"
            title="Open Developer Settings"
            type="button"
          >
            <span
              aria-label="OAuth App"
              class="mr-1"
            >
              <svg
                aria-hidden="true"
                class="octicon octicon-person"
                fill="currentColor"
                focusable="false"
                height="12"
                style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
                viewBox="0 0 16 16"
                width="12"
              >
                <path
                  d="M10.561 8.073a6.005 6.005 0 0 1 3.432 5.142.75.75 0 1 1-1.498.07 4.5 4.5 0 0 0-8.99 0 .75.75 0 0 1-1.498-.07 6.004 6.004 0 0 1 3.431-5.142 3.999 3.999 0 1 1 5.123 0ZM10.5 5a2.5 2.5 0 1 0-5 0 2.5 2.5 0 0 0 5 0Z"
                />
              </svg>
            </span>
            OAuth App
          </button>
        </div>
        <div>
          <button
            class="hover:text-gray-500 py-1 px-2 my-1 mx-2 focus:outline-none"
            title="Logout octocat"
            type="button"
          >
            <svg
              aria-label="Logout octocat"
              class="octicon octicon-sign-out"
              fill="currentColor"
              focusable="false"
              height="20"
              role="img"
              style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
              viewBox="0 0 16 16"
              width="20"
            >
              <path
                d="M2 2.75C2 1.784 2.784 1 3.75 1h2.5a.75.75 0 0 1 0 1.5h-2.5a.25.25 0 0 0-.25.25v10.5c0 .138.112.25.25.25h2.5a.75.75 0 0 1 0 1.5h-2.5A1.75 1.75 0 0 1 2 13.25Zm10.44 4.5-1.97-1.97a.749.749 0 0 1 .326-1.275.749.749 0 0 1 .734.215l3.25 3.25a.75.75 0 0 1 0 1.06l-3.25 3.25a.749.749 0 0 1-1.275-.326.749.749 0 0 1 .215-.734l1.97-1.97H6.75a.75.75 0 0 1 0-1.5Z"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="flex justify-between items-center bg-gray-100 dark:bg-gray-900 rounded-md p-1 mb-4"
      >
        <div
          class="ml-2 text-xs"
        >
          <button
            class="cursor-pointer font-semibold mb-1 text-sm"
            title="Open Profile"
            type="button"
          >
            @
            octocat
            <span
              class="pl-1 font-medium text-xs italic"
            >
              - 
              Mona Lisa Octocat
            </span>
          </button>
          <button
            class="cursor-pointer mb-1 ml-1 align-middle"
            title="Open Host"
            type="button"
          >
            <span
              aria-label="GitHub Cloud"
              class="mr-1"
            >
              <svg
                aria-hidden="true"
                class="octicon octicon-mark-github"
                fill="currentColor"
                focusable="false"
                height="12"
                style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
                viewBox="0 0 16 16"
                width="12"
              >
                <path
                  d="M8 0c4.42 0 8 3.58 8 8a8.013 8.013 0 0 1-5.45 7.59c-.4.08-.55-.17-.55-.38 0-.27.01-1.13.01-2.2 0-.75-.25-1.23-.54-1.48 1.78-.2 3.65-.88 3.65-3.95 0-.88-.31-1.59-.82-2.15.08-.2.36-1.02-.08-2.12 0 0-.67-.22-2.2.82-.64-.18-1.32-.27-2-.27-.68 0-1.36.09-2 .27-1.53-1.03-2.2-.82-2.2-.82-.44 1.1-.16 1.92-.08 2.12-.51.56-.82 1.28-.82 2.15 0 3.06 1.86 3.75 3.64 3.95-.23.2-.44.55-.51 1.07-.46.21-1.61.55-2.33-.66-.15-.24-.6-.83-1.23-.82-.67.01-.27.38.01.53.34.19.73.9.82 1.13.16.45.68 1.31 2.69.94 0 .67.01 1.3.01 1.49 0 .21-.15.45-.55.38A7.995 7.995 0 0 1 0 8c0-4.42 3.58-8 8-8Z"
                />
              </svg>
            </span>
            GitHub Cloud
             - 
            github.com
          </button>
          <button
            class="cursor-pointer ml-1 align-middle"
            title="Open Developer Settings"
            type="button"
          >
            <span
              aria-label="GitHub App"
              class="mr-1"
            >
              <svg
                aria-hidden="true"
                class="octicon octicon-apps"
                fill="currentColor"
                focusable="false"
                height="12"
                style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
                viewBox="0 0 16 16"
                width="12"
              >
                <path
                  d="M1.5 3.25c0-.966.784-1.75 1.75-1.75h2.5c.966 0 1.75.784 1.75 1.75v2.5A1.75 1.75 0 0 1 5.75 7.5h-2.5A1.75 1.75 0 0 1 1.5 5.75Zm7 0c0-.966.784-1.75 1.75-1.75h2.5c.966 0 1.75.784 1.75 1.75v2.5a1.75 1.75 0 0 1-1.75 1.75h-2.5A1.75 1.75 0 0 1 8.5 5.75Zm-7 7c0-.966.784-1.75 1.75-1.75h2.5c.966 0 1.75.784 1.75 1.75v2.5a1.75 1.75 0 0 1-1.75 1.75h-2.5a1.75 1.75 0 0 1-1.75-1.75Zm7 0c0-.966.784-1.75 1.75-1.75h2.5c.966 0 1.75.784 1.75 1.75v2.5a1.75 1.75 0 0 1-1.75 1.75h-2.5a1.75 1.75 0 0 1-1.75-1.75ZM3.25 3a.25.25 0 0 0-.25.25v2.5c0 .138.112.25.25.25h2.5A.25.25 0 0 0 6 5.75v-2.5A.25.25 0 0 0 5.75 3Zm7 0a.25.25 0 0 0-.25.25v2.5c0 .138.112.25.25.25h2.5a.25.25 0 0 0 .25-.25v-2.5a.25.25 0 0 0-.25-.25Zm-7 7a.25.25 0 0 0-.25.25v2.5c0 .138.112.25.25.25h2.5a.25.25 0 0 0 .25-.25v-2.5a.25.25 0 0 0-.25-.25Zm7 0a.25.25 0 0 0-.25.25v2.5c0 .138.112.25.25.25h2.5a.25.25 0 0 0 .25-.25v-2.5a.25.25 0 0 0-.25-.25Z"
                />
              </svg>
            </span>
            GitHub App
          </button>
        </div>
        <div>
          <button
            class="hover:text-gray-500 py-1 px-2 my-1 mx-2 focus:outline-none"
            title="Logout octocat"
            type="button"
          >
            <svg
              aria-label="Logout octocat"
              class="octicon octicon-sign-out"
              fill="currentColor"
              focusable="false"
              height="20"
              role="img"
              style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
              viewBox="0 0 16 16"
              width="20"
            >
              <path
                d="M2 2.75C2 1.784 2.784 1 3.75 1h2.5a.75.75 0 0 1 0 1.5h-2.5a.25.25 0 0 0-.25.25v10.5c0 .138.112.25.25.25h2.5a.75.75 0 0 1 0 1.5h-2.5A1.75 1.75 0 0 1 2 13.25Zm10.44 4.5-1.97-1.97a.749.749 0 0 1 .326-1.275.749.749 0 0 1 .734.215l3.25 3.25a.75.75 0 0 1 0 1.06l-3.25 3.25a.749.749 0 0 1-1.275-.326.749.749 0 0 1 .215-.734l1.97-1.97H6.75a.75.75 0 0 1 0-1.5Z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="flex justify-between items-center bg-gray-200 dark:bg-gray-darker py-1 px-8 text-sm"
  >
    <div
      class="font-semibold italic"
    >
      Add new account
    </div>
    <div>
      <button
        class="hover:text-gray-500 py-1 px-2 my-1 mx-2 focus:outline-none"
        hidden=""
        title="Login with Personal Access Token"
        type="button"
      >
        <svg
          aria-label="Login with Personal Access Token"
          class="octicon octicon-key"
          fill="currentColor"
          focusable="false"
          height="18"
          role="img"
          style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
          viewBox="0 0 16 16"
          width="18"
        >
          <path
            d="M10.5 0a5.499 5.499 0 1 1-1.288 10.848l-.932.932a.749.749 0 0 1-.53.22H7v.75a.749.749 0 0 1-.22.53l-.5.5a.749.749 0 0 1-.53.22H5v.75a.749.749 0 0 1-.22.53l-.5.5a.749.749 0 0 1-.53.22h-2A1.75 1.75 0 0 1 0 14.25v-2c0-.199.079-.389.22-.53l4.932-4.932A5.5 5.5 0 0 1 10.5 0Zm-4 5.5c-.001.431.069.86.205 1.269a.75.75 0 0 1-.181.768L1.5 12.56v1.69c0 .138.112.25.25.25h1.69l.06-.06v-1.19a.75.75 0 0 1 .75-.75h1.19l.06-.06v-1.19a.75.75 0 0 1 .75-.75h1.19l1.023-1.025a.75.75 0 0 1 .768-.18A4 4 0 1 0 6.5 5.5ZM11 6a1 1 0 1 1 0-2 1 1 0 0 1 0 2Z"
          />
        </svg>
        <svg
          aria-hidden="true"
          class="ml-1 mb-2"
          fill="currentColor"
          focusable="false"
          height="10"
          style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
          viewBox="0 0 16 16"
          width="10"
        >
          <path
            d="M7.75 2a.75.75 0 0 1 .75.75V7h4.25a.75.75 0 0 1 0 1.5H8.5v4.25a.75.75 0 0 1-1.5 0V8.5H2.75a.75.75 0 0 1 0-1.5H7V2.75A.75.75 0 0 1 7.75 2Z"
          />
        </svg>
      </button>
      <button
        class="hover:text-gray-500 py-1 px-2 my-1 mx-2 focus:outline-none"
        hidden=""
        title="Login with OAuth App"
        type="button"
      >
        <svg
          aria-label="Login with OAuth App"
          class="octicon octicon-person"
          fill="currentColor"
          focusable="false"
          height="20"
          role="img"
          style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
          viewBox="0 0 16 16"
          width="20"
        >
          <path
            d="M10.561 8.073a6.005 6.005 0 0 1 3.432 5.142.75.75 0 1 1-1.498.07 4.5 4.5 0 0 0-8.99 0 .75.75 0 0 1-1.498-.07 6.004 6.004 0 0 1 3.431-5.142 3.999 3.999 0 1 1 5.123 0ZM10.5 5a2.5 2.5 0 1 0-5 0 2.5 2.5 0 0 0 5 0Z"
          />
        </svg>
        <svg
          aria-hidden="true"
          class="mb-2"
          fill="currentColor"
          focusable="false"
          height="10"
          style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
          viewBox="0 0 16 16"
          width="10"
        >
          <path
            d="M7.75 2a.75.75 0 0 1 .75.75V7h4.25a.75.75 0 0 1 0 1.5H8.5v4.25a.75.75 0 0 1-1.5 0V8.5H2.75a.75.75 0 0 1 0-1.5H7V2.75A.75.75 0 0 1 7.75 2Z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
